src/vs/workbench/contrib/terminal/browser/terminalExtensions.ts

```diff
@@ -5,19 +5,22 @@
 
 import { BrandedService, IConstructorSignature } from 'vs/platform/instantiation/common/instantiation';
 import { Registry } from 'vs/platform/registry/common/platform';
-import { ITerminalContribution, ITerminalInstance } from 'vs/workbench/contrib/terminal/browser/terminal';
+import { IDetachedTerminalInstance, ITerminalContribution, ITerminalInstance } from 'vs/workbench/contrib/terminal/browser/terminal';
 import { TerminalWidgetManager } from 'vs/workbench/contrib/terminal/browser/widgets/widgetManager';
-import { ITerminalProcessManager } from 'vs/workbench/contrib/terminal/common/terminal';
+import { ITerminalProcessInfo, ITerminalProcessManager } from 'vs/workbench/contrib/terminal/common/terminal';
 
 export type TerminalContributionCtor = IConstructorSignature<ITerminalContribution, [ITerminalInstance, ITerminalProcessManager, TerminalWidgetManager]>;
+export type DetachedTerminalContributionCtor = IConstructorSignature<ITerminalContribution, [IDetachedTerminalInstance, ITerminalProcessInfo, TerminalWidgetManager]>;
 
-export interface ITerminalContributionDescription {
-	readonly id: string;
-	readonly ctor: TerminalContributionCtor;
-}
+export type ITerminalContributionDescription = { readonly id: string } & (
+	| { readonly canRunInDetachedTerminals: false; readonly ctor: TerminalContributionCtor | DetachedTerminalContributionCtor }
+	| { readonly canRunInDetachedTerminals: true; readonly ctor: DetachedTerminalContributionCtor }
+);
```

Comment:
I was expecting this?

```ts
	| { readonly canRunInDetachedTerminals: false; readonly ctor: TerminalContributionCtor }
	| { readonly canRunInDetachedTerminals: true; readonly ctor: TerminalContributionCtor 
 | DetachedTerminalContributionCtor }
```

Reply:
I think I understand what you're going for, but it's pretty confusing to look at and took me a while to understand:

```ts
canRunInDetachedTerminals: false; readonly ctor: ... | DetachedTerminalContributionCtor
```

How about this?

```ts
export type TerminalContributionCtor = IConstructorSignature<ITerminalContribution, [ITerminalInstance, ITerminalProcessManager, TerminalWidgetManager]>;
export type DetachedCompatibleTerminalContributionCtor = IConstructorSignature<ITerminalContribution, [IDetachedTerminalInstance, ITerminalProcessInfo, TerminalWidgetManager]>;

export type ITerminalContributionDescription = { readonly id: string } & (
	| { readonly canRunInDetachedTerminals: false; readonly ctor: TerminalContributionCtor }
	| { readonly canRunInDetachedTerminals: true; readonly ctor: DetachedCompatibleTerminalContributionCtor }
```

Reply:
yep, that works